zephyr/lib/regions_mm.c

```diff
@@ -1,52 +1,759 @@
+	list_item_append(&vmh_list, &new_heap->node);
+
+	return new_heap;
+fail:
+	for (i = 0; i < MAX_MEMORY_ALLOCATORS_COUNT; i++) {
+		struct sys_mem_blocks *allocator = new_heap->physical_blocks_allocators[i];
+		struct sys_bitarray *allocation_sizes = new_heap->allocation_sizes[i];
+
+		if (allocator) {
+			if (allocator->bitmap)
+				rfree(allocator->bitmap->bundles);
+			rfree(allocator->bitmap);
+		}
+		if (allocation_sizes) {
+			if (allocation_sizes->bundles)
+				rfree(allocation_sizes->bundles);
+			rfree(allocation_sizes);
+		}
+		rfree(allocator);
+	}
+	rfree(new_heap);
+	return NULL;
+}
+
+/**
+ * @brief Alloc function
+ *
+ * Allocates memory on heap from provided heap pointer.
+ * Check if we need to map physical memory for given allocation
+ * and do it when necessary.
+ * @param heap pointer to a heap that will be used for allocation.
+ * @param alloc_size Size in bytes to allocate.
+ *
+ * @retval ptr to allocated memory if successful
+ * @retval NULL on allocation failure
+ */
+void *vmh_alloc(struct vmh_heap *heap, uint32_t alloc_size)
+{
+	if (!alloc_size)
+		return NULL;
+	/* Only operations on the same core are allowed */
+	if (heap->core_id != cpu_get_id())
+		return NULL;
+
+	void *ptr = NULL;
+	uintptr_t phys_aligned_ptr, phys_aligned_alloc_end;
+	uint32_t block_size, allocation_error_code = -ENOMEM;
+	size_t i, mem_block_iterator, allocation_bitarray_offset, allocation_bitarray_position,
+		check_offset, check_size, check_position, block_count, physical_block_count;
+	uintptr_t phys_block_ptr, phys_block_end, block_realigned_ptr,
+		block_realigned_alloc_end;
+	uintptr_t *ptrs_to_map = NULL;
+
+	/* We will gather error code when allocating on physical block
+	 * allocators.
+	 * Loop will continue as long as there are blocks to iterate on and
+	 * if allocation keeps returning error code.
+	 * When alloc succeeded it will return 0 on allocation and we break
+	 * form loop and continue.
+	 * If allocating_continuously is on for heap, allocation will be made on first available
+	 * block series on given block bundle. For allocating_continuously off allocation will only
+	 * happened on first free block on first allocator that can accommodate
+	 * requested size.
+	 */
+	for (mem_block_iterator = 0; mem_block_iterator < MAX_MEMORY_ALLOCATORS_COUNT;
+	     mem_block_iterator++) {
+
+		/* continiue so we do not check mem blocks that do not exist */
+		if (!heap->physical_blocks_allocators[mem_block_iterator])
+			continue;
+
+		/* calculate block count needed to allocate for current
+		 * mem_block.
+		 */
+		block_size =
+		    1 << heap->physical_blocks_allocators[mem_block_iterator]->blk_sz_shift;
+		block_count = SOF_DIV_ROUND_UP((uint64_t)alloc_size, (uint64_t)block_size);
+
+		if (block_count >
+		    heap->physical_blocks_allocators[mem_block_iterator]->num_blocks)
+			continue;
+		/* Try span alloc on first available mem_block for non span
+		 * check if block size is sufficient.
+		 */
```

Comment:
This has nothing to do with contiguous physical pages. 
This is firstly checking if we have virtual blocks of sufficient size by allocating on memblocks. If allocation was successful    we allocate physical pages if needed.
And the answer is we do not need it and I do not force anything like that here.

Reply:
ah, ok, so we just map them into a contiguous address region, but physical pages can be discontiguous. Maybe it would be good to clarify that somewhere in a comment

Reply:
When this gets in, one of the next steps would be documenting it. I think adding anything about span physical mapping would only cause confusion here since there is a clear distinction in wording - allocation / mapping.